### ISSUE TITLE
Trying to replace only k/v in a hash fails validation

### ISSUE THREAD

**Issue by TJM** [xvik/yaml-updater#11]
We have found that if you try to replace a key/value that is the only item in the hash, it fails. The output indicates that it did exactly what I wanted, but something in the validation steps goes wrong.

Here is a simple example:

* source.yaml:

```yaml
test: value
spec:
    image: probably
```

* merge.yaml:

```yaml
 spec:
  image: absolutely
```

* output:
```console

bash-5.1# java -jar /usr/local/bin/yaml-config-updater.jar source.yaml merge.yaml -d spec.image --verbose
Updating configuration: /source.yaml
[main] DEBUG ru.vyarus.yaml.updater.YamlUpdater - Parsing new configuration...
[main] DEBUG ru.vyarus.yaml.updater.update.EnvSupport - Replacing variables in format '#{name}' from:
    PATH=/opt/openjdk-18/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
    HOSTNAME=8a245a7c45c6
    LD_LIBRARY_PATH=/opt/openjdk-18/lib/server:/opt/openjdk-18/lib:/opt/openjdk-18/../lib
    JAVA_HOME=/opt/openjdk-18
    TERM=xterm
    JAVA_VERSION=18-ea+11
    PWD=/
    SHLVL=1
    HOME=/root
    _=/opt/openjdk-18/bin/java
[main] INFO ru.vyarus.yaml.updater.YamlUpdater - Environment variables applied to new config
[main] INFO ru.vyarus.yaml.updater.YamlUpdater - New configuration parsed (27 bytes, 2 lines)
[main] DEBUG ru.vyarus.yaml.updater.YamlUpdater - Parsing current configuration file (/source.yaml)...
[main] INFO ru.vyarus.yaml.updater.YamlUpdater - Removing configuration property: spec/image
[main] INFO ru.vyarus.yaml.updater.YamlUpdater - Current configuration parsed (38 bytes, 3 lines)
[main] DEBUG ru.vyarus.yaml.updater.YamlUpdater - Merging configurations...
[main] INFO ru.vyarus.yaml.updater.YamlUpdater - Configuration merged
[main] DEBUG ru.vyarus.yaml.updater.YamlUpdater - Validating merged result
java.lang.IllegalStateException: Failed to update: original configuration remains
	at ru.vyarus.yaml.updater.YamlUpdater.execute(YamlUpdater.java:162)
	at ru.vyarus.yaml.updater.UpdateConfig$Configurator.update(UpdateConfig.java:320)
	at ru.vyarus.yaml.updater.cli.UpdateConfigCli.call(UpdateConfigCli.java:93)
	at ru.vyarus.yaml.updater.cli.UpdateConfigCli.call(UpdateConfigCli.java:26)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1953)
	at picocli.CommandLine.access$1300(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2352)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2346)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2311)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
	at picocli.CommandLine.execute(CommandLine.java:2078)
	at ru.vyarus.yaml.updater.cli.UpdateConfigCli.main(UpdateConfigCli.java:111)
Caused by: java.lang.IllegalStateException: Failed to validate merge result:

   1| test: value
   2| spec:
   3|   image: absolutely

	at ru.vyarus.yaml.updater.YamlUpdater.validateResult(YamlUpdater.java:332)
	at ru.vyarus.yaml.updater.YamlUpdater.execute(YamlUpdater.java:159)
	... 11 more
Caused by: java.lang.IllegalStateException: Value 'spec' disappeared (should remain from original file): 'null'
	at ru.vyarus.yaml.updater.update.UpdateResultValidator.checkMissedValues(UpdateResultValidator.java:114)
	at ru.vyarus.yaml.updater.update.UpdateResultValidator.validate(UpdateResultValidator.java:57)
	at ru.vyarus.yaml.updater.YamlUpdater.validateResult(YamlUpdater.java:305)
	... 12 more

```

... it does work fine, if there is another tag at the same level.... for example:

* source.yaml:

```yaml
test: value
spec:
    image: probably
    something: else
```

* output:

```console
bash-5.1# java -jar /usr/local/bin/yaml-config-updater.jar source.yaml merge.yaml -d spec.image --verbose
Updating configuration: /source.yaml
[main] DEBUG ru.vyarus.yaml.updater.YamlUpdater - Parsing new configuration...
[main] DEBUG ru.vyarus.yaml.updater.update.EnvSupport - Replacing variables in format '#{name}' from:
    PATH=/opt/openjdk-18/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
    HOSTNAME=8a245a7c45c6
    LD_LIBRARY_PATH=/opt/openjdk-18/lib/server:/opt/openjdk-18/lib:/opt/openjdk-18/../lib
    JAVA_HOME=/opt/openjdk-18
    TERM=xterm
    JAVA_VERSION=18-ea+11
    PWD=/
    SHLVL=1
    HOME=/root
    _=/opt/openjdk-18/bin/java
[main] INFO ru.vyarus.yaml.updater.YamlUpdater - Environment variables applied to new config
[main] INFO ru.vyarus.yaml.updater.YamlUpdater - New configuration parsed (27 bytes, 2 lines)
[main] DEBUG ru.vyarus.yaml.updater.YamlUpdater - Parsing current configuration file (/source.yaml)...
[main] INFO ru.vyarus.yaml.updater.YamlUpdater - Removing configuration property: spec/image
[main] INFO ru.vyarus.yaml.updater.YamlUpdater - Current configuration parsed (58 bytes, 4 lines)
[main] DEBUG ru.vyarus.yaml.updater.YamlUpdater - Merging configurations...
[main] INFO ru.vyarus.yaml.updater.YamlUpdater - Configuration merged
[main] DEBUG ru.vyarus.yaml.updater.YamlUpdater - Validating merged result
[main] INFO ru.vyarus.yaml.updater.YamlUpdater - Merged file correctness validated
[main] INFO ru.vyarus.yaml.updater.YamlUpdater - Backup created: /source.yaml.20211109230832
[main] INFO ru.vyarus.yaml.updater.YamlUpdater - Configuration updated: /source.yaml

Configuration: /source.yaml (58 bytes, 4 lines)
Updated from source of 27 bytes, 2 lines
Resulted in 56 bytes, 4 lines

	Removed from old file:
		spec/image                               3  | image: probably

	Added from new file:
		spec/image                               2  | image: absolutely

	Backup created: source.yaml.20211109230832

bash-5.1#
```

**Comment by xvik** [comment]
Thank you very much for the detailed report! A new version would be released soon: I want to try adding `--override` option for overriding mode first (for patching existing yaml configs - your case).

Remembering your first issue I thought: maybe you can use [variables](https://github.com/xvik/yaml-updater/tree/master/yaml-config-updater-cli#variables) instead of "patching" yaml file. Of course, this is possible only in case if you control the original yaml source. In this case, you can use placeholders:

(template.yaml)
```yaml
....

spec:
  values:
    image: #{imageType}
```

And allow the tool to put desired value for you:

```
java -jar /usr/local/bin/yaml-config-updater.jar source.yaml template.yaml -e imageType=absolutely
```

If source.yaml file does not exist, the tool would process variables in the template and put the result as the source.yaml file.

**Comment by TJM** [comment]
I could use a variable in the "change.yaml" (template.yaml) file, but since I am going to "generate" the file on the fly[1] anyhow, I might as well just put the value in, rather than a variable reference.

Would `--override` just allow me to override the value without the `--delete spec.vaules.image.tag` (thus maybe keeping the order of the keys the same)?

.. or would I be able to do something like `--override spec.values.image.tag=theValue` <-- that would obviate the need for the change.yaml file in my case, but my case (changing one value) might be so limited that it is not really worth messing around with.

[1]: Generate change.yaml:
```bash
echo -e "spec:\n  values:\n    image:\n      tag: ${RELEASE_TAG}\n" > change.yaml
```
... I could replace the `$` with `#`, but the value add would be minimal. I could also use a heredoc, but again, its only 3 lines.

**Comment by TJM** [comment]
Also, thanks for the quick fix!

**Comment by xvik** [comment]
> ... I could replace the $ with #, but the value add would be minimal. I could also use a heredoc, but again, its only 4 lines.

Agree, in your case variables would be useless.

> Would --override just allow me to override the value without the --delete spec.vaules.image.tag (thus maybe keeping the order of the keys the same)?

Yes, I was thinking about the "patch" use case when some existing configuration values must be changed during migration and small migration.yaml would be prepared containing only values that must be changed (and, possibly, values to add too). In general, `--override` option would mean overriding existing values instead of not touching. 

But I don't promise that it would be eventually done: I'm still trying to figure out if there is a real need for such usage. 
For your case changing `--delete spec.vaules.image.tag`  to `--override` would lead to exactly the same result. 
In general case, deletion + addition may change the order of properties if old config contain more values then new one, so `--override` approach would be more consistent.

In any case, new version would be released soon.

**Comment by TJM** [comment]
I think, maybe even `--allow-override` or `--allow-overwrite` in lieu of deleting the specific node? Maybe we should split this to its own issue? :)

**Comment by xvik** [comment]
`allow-override` assumes there is a possibility to not overwrite value (tool should decide), but it is not the case.
I like `--overwrite` name - semantically it's more correct.

> Maybe we should split this to its own issue?

No need, I already looking it and wouldn't forget.

**Comment by TJM** [comment]
Ah, I did not know --overwrite would still require the attribute name. I was thinking of a general "allow any overwrites" type option. Thanks!

**Comment by xvik** [comment]
Released 1.3.3 fixing the issue.

Overwrite option is postponed: technically I did it, but the problem is in object lists (they are always a problem) - it's not clear how it should behave on them (more correctly, all currently possible behaviors are clumsy). I'll wait until the current object lists merging behavior would be stabilized by practice.
